File: 백준/Silver/10971. 외판원 순회 2/외판원 순회 2.py
import sys
n = int(sys.stdin.readline())

map_city = [list(map(int,sys.stdin.readline().split())) for _ in range(n)]

visited = [False] * n
min_cost = float("INF")

# 모든 depth를 한 번씩만 방문 / 출발도시로 돌아오는 경로 고려
def dfs(city, path, cost):
    global min_cost

    # 모든 도시를 방문했을 경우
    if len(path) == n:
        # 출발 도시로 돌아가는 비용 추가
        return_cost = map_city[city][path[0]]
        if return_cost > 0:
            min_cost = min(min_cost, cost + return_cost)
        return

    for next_city in range(n):
        # 방문하지 않았고, 이동할 수 있는 경우
        if not visited[next_city] and map_city[city][next_city] > 0:
            visited[next_city] = True
            dfs(next_city, path + [next_city], cost + map_city[city][next_city])
            visited[next_city] = False


# 각 도시마다 방문
for i in range(n):
    visited[i] = True
    dfs(i,[i],0)
    visited[i] = False

print(min_cost)

File: 백준/Silver/10971. 외판원 순회 2/외판원 순회 2.java
import java.io.*;

class Main {
  static int[][] w;
  static boolean[] visited;
  static int N;
  static int min = Integer.MAX_VALUE;
   public static void main(String[] args) throws IOException {
      BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

      N = Integer.parseInt(br.readLine());

      w = new int[N][N];

      for (int i = 0; i < N; i++) {
        String[] st = br.readLine().split(" ");

        for (int j = 0; j < N; j++) {
          w[i][j] = Integer.parseInt(st[j]);
        }

      }

      visited = new boolean[N];
      visited[0] = true;

      dfs(0,  0, 0, 1);

      System.out.println(min);
   }
   static void dfs(int start, int now, int cost, int cnt) {
    if (cnt == N) {
      // 모두 방문한 경우
      if (w[now][start] != 0) {
        min = Math.min(min, cost + w[now][start]);
      }
      return;
    }


    for (int next = 0; next < N; next++) {
      if (!visited[next] && w[now][next] != 0) {
        visited[next] = true;
        dfs(start, next, cost + w[now][next], cnt+1);
        visited[next] = false;
      }
    }
   }
}